homeassistant/components/melcloud/water_heater.py

```diff
@@ -0,0 +1,133 @@
+"""Platform for water_heater integration."""
+from typing import List, Optional
+
+from pymelcloud import DEVICE_TYPE_ATW, AtwDevice
+from pymelcloud.atw_device import (
+    PROPERTY_OPERATION_MODE,
+    PROPERTY_TARGET_TANK_TEMPERATURE,
+)
+from pymelcloud.device import PROPERTY_POWER
+
+from homeassistant.components.water_heater import (
+    SUPPORT_OPERATION_MODE,
+    SUPPORT_TARGET_TEMPERATURE,
+    WaterHeaterDevice,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import TEMP_CELSIUS
+from homeassistant.helpers.typing import HomeAssistantType
+
+from . import DOMAIN, MelCloudDevice
+from .const import ATTR_STATUS, TEMP_UNIT_LOOKUP
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+):
+    """Set up MelCloud device climate based on config_entry."""
+    mel_devices = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        [
+            AtwWaterHeater(mel_device, mel_device.device)
+            for mel_device in mel_devices[DEVICE_TYPE_ATW]
+        ],
+        True,
+    )
+
+
+class AtwWaterHeater(WaterHeaterDevice):
+    """Air-to-Water water heater."""
+
+    def __init__(self, api: MelCloudDevice, device: AtwDevice):
+        """Initialize water heater device."""
+        self._api = api
+        self._device = device
+        self._name = device.name
+
+    async def async_update(self):
+        """Update state from MELCloud."""
+        await self._api.async_update()
+
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return a unique ID."""
+        return f"{self._api.device.serial}-{self._api.device.mac}"
+
+    @property
+    def name(self):
+        """Return the display name of this entity."""
+        return self._name
+
+    @property
+    def device_info(self):
+        """Return a device description for device registry."""
+        return self._api.device_info
```

Comment:
Does this include a connection for the mac ?

Reply:
I honestly don't know what the connections should be in the `device_info`. The mac is used only as device identitier and a redundant one at that.

What I am missing is are the explicit `via_device` entries to tie the zone climates to the `water_heater` and the sensors to correct devices. I don't know if it is necessary tho as the entities are already linked correctly to the device.

Reply:
By specifying the mac in device info it will be able to work together with other integrations in monitoring this device. Example https://github.com/home-assistant/core/blob/dev/homeassistant/components/daikin/__init__.py#L147